core/txpool/legacypool/celo_list.go

```diff
@@ -0,0 +1,170 @@
+package legacypool
+
+import (
+	"math"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/contracts"
+	"github.com/ethereum/go-ethereum/core/txpool"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+func (l *list) FilterWhitelisted(rates common.ExchangeRates) (types.Transactions, types.Transactions) {
+	removed := l.txs.Filter(func(tx *types.Transaction) bool {
+		return !common.IsCurrencyWhitelisted(rates, tx.FeeCurrency())
+	})
+
+	if len(removed) == 0 {
+		return nil, nil
+	}
+
+	invalid := l.dropInvalidsAfterRemovalAndReheap(removed)
+	l.subTotalCost(removed)
+	l.subTotalCost(invalid)
+	return removed, invalid
+}
+
+func (l *list) dropInvalidsAfterRemovalAndReheap(removed types.Transactions) types.Transactions {
+	var invalids types.Transactions
+	// If the list was strict, filter anything above the lowest nonce
+	// Note that the 'invalid' txs have no intersection with the 'removed' txs
+	if l.strict {
+		lowest := uint64(math.MaxUint64)
+		for _, tx := range removed {
+			if nonce := tx.Nonce(); lowest > nonce {
+				lowest = nonce
+			}
+		}
+		invalids = l.txs.filter(func(tx *types.Transaction) bool { return tx.Nonce() > lowest })
+	}
+	l.txs.reheap()
+	return invalids
+}
+
+func (l *list) FeeCurrencies() []common.Address {
+	currencySet := make(map[common.Address]interface{})
+	for _, tx := range l.txs.items {
+		// native currency (nil) represented as Zero address
+		currencySet[getCurrencyKey(tx.FeeCurrency())] = struct{}{}
+	}
+	currencies := make([]common.Address, 0, len(currencySet))
+	for curr := range currencySet {
+		currencies = append(currencies, curr)
+	}
+	return currencies
+}
+
+func getCurrencyKey(feeCurrency *common.Address) common.Address {
+	if feeCurrency == nil {
+		return common.ZeroAddress
+	}
+	return *feeCurrency
+}
+
+func (c *list) totalCostVar(feeCurrency *common.Address) *big.Int {
+	key := getCurrencyKey(feeCurrency)
+	if tc, ok := c.totalCost[key]; ok {
+		return tc
+	}
+	newTc := big.NewInt(0)
+	c.totalCost[key] = newTc
+	return newTc
+}
+
+func (c *list) TotalCostFor(feeCurrency *common.Address) *big.Int {
+	if tc, ok := c.totalCost[getCurrencyKey(feeCurrency)]; ok {
+		return new(big.Int).Set(tc)
+	}
+	return big.NewInt(0)
+}
+
+func (c *list) costCapFor(feeCurrency *common.Address) *big.Int {
+	if tc, ok := c.costCap[getCurrencyKey(feeCurrency)]; ok {
```

Comment:
since costCap is not an accumulator, but only the max between(previous cap, max tx cost in list), I thought it looked better to just get the total cap, and, if a tx is accepted, just call to 'updateCap' which may or may not update it depending on the comparison